library/bignum_core.h

```diff
@@ -123,4 +123,37 @@ mbedtls_mpi_uint mbedtls_mpi_core_add_if( mbedtls_mpi_uint *d,
                                           size_t n,
                                           unsigned cond );
 
+/**
+ * \brief Calculate initialisation value for fast Montgomery modular
+ *        multiplication
+ *
+ * \param mm   Pointer to an mbedtls_mpi_uint to receive the result.
+ * \param N    Little-endian presentation of the modulus, which must be odd.
+ */
+void mbedtls_mpi_montg_init( mbedtls_mpi_uint *mm, const mbedtls_mpi *N );
+
+/** Montgomery multiplication: A = A * B * R^-1 mod N  (HAC 14.36)
+ *
+ * \param[in,out]   A   One of the numbers to multiply.
+ *                      It must have at least as many limbs as N
+ *                      (A->n >= N->n), and any limbs beyond n are ignored.
+ *                      On successful completion, A contains the result of
+ *                      the multiplication A * B * R^-1 mod N where
+ *                      R = (2^ciL)^n.
+ * \param[in]       B   One of the numbers to multiply.
+ *                      It must be nonzero and must not have more limbs than N
+ *                      (B->n <= N->n).
+ * \param[in]       N   The modulo. N must be odd.
+ * \param           mm  The value calculated by `mbedtls_mpi_montg_init(&mm, N)`.
+ *                      This is -N^-1 mod 2^ciL.
+ * \param[in,out]   T   A bignum for temporary storage.
+ *                      It must be at least twice the limb size of N plus 1
+ *                      (T->n >= 2 * N->n + 1).
+ *                      Its initial content is unused and
+ *                      its final content is indeterminate.
+ *                      Note that unlike the usual convention in the library
+ *                      for `const mbedtls_mpi*`, the content of T can change.
```

Comment:
Instead of documenting this exception, wouldn't it be better to maintain the convention and remove the `const` from the `T` parameter?

Reply:
This is the documentation for the existing `mpi_montmul()` function, brought over as-is from the current implementation. This doc was added in PR 3398 by Gilles, and was reviewed by you and mpg :)

And: I don't know if there is a good reason why `T` was originally `const`, and then documented as such.

As a matter of principle for this work, to what extent do we want to tidy up the existing code? It makes it harder to search for things in a commit if they are removed from one file and put into another, but also edited at the same time. And creating separate commits for the moving/editing is going to slow down bringing in the new bignum code from the prototype...

Reply:
Indeed, it is pre-existing. (Gilles's commit was purely about documentation and already beyond the strict scope of the PR. Even if that wasn't the case, however hard we try, we miss things. Also, I would like to think that all of us got better in these two years that has past since.)

I don't see any good reason, that `const` qualifier is ancient, it was added in this commit: ff60ee6c2a. If there is a good reason to it, I would think that the compiler will let us know immediately.

I don't think that we should spend too much time on tidying up existing code. That said, we shouldn't leave everything as it is either. (You know, merciless refactoring and always leaving the code in better shape than we found it and things like that.)

In summary, I think that we can go for a low hanging fruit here and there, especially if it is about the consistency of the new part of the Bignum module.

In this specific case, we don't need to fix it here and now if you don't feel like it.

(Regarding the prototype code, I wouldn't consider it existing code. It is new code that we are bringing in, we trust the correctness (until tests prove otherwise), but improving style and consistency should be within scope.)

Creating separate commits for editing adds some time, but compared to the rest of the work (writing the tests and new code, thinking and having the discussions about it) I would imagine it would be very low cost.

Reply:
[Last time](https://github.com/Mbed-TLS/mbedtls/pull/3398/commits/2a82f72703c3cbb90697c13bdf57184c0547b653#r435737663) we decided to leave the `const` there. It's a rare case in the bignum code where we want the middle ground between a read-only MPI and a modifiable MPI: an MPI where the limbs may be modified but which may not be resized. I think it made sense to keep the `const` then. I don't know if it still makes sense now.

As noted in the comment, here we're using `const` for exactly what it means in C. In most other places with `const mbedtls_mpi*`, we're using `const` to mean more than what it means in C.

Reply:
In that case we should remove the `const` as the context has changed: functions in this file operate on fixed size MPIs and never reallocate, there is no need to signal that it won't be reallocated.

Reply:
Oh, sorry, I can see that it has been removed at a later commit.

Reply:
Actually, that commit just puts the doc comments back where they came from, unchanged - that function became public briefly so it could be tested, but then put back to `static` when it was modified to call the new function.

I have added a commit that removes the `const` qualifier.